### DISCUSSION TITLE
PDFsharp 6.2 Preview 1 was released

### DISCUSSION THREAD

**Discussion by ThomasHoevel** [empira/PDFsharp#150] [Announcements]
More information here:  
https://docs.pdfsharp.net/General/History.html#ps-6-2-0-pv1

**Comment by ZephyrZiggurat** [comment]
Regarding the upcoming digital signatures, is/will it possible to add a blank signature field? I have an existing pdf with basically just text on it (no form) and I'd like to be able to drop a blank digital signature field somewhere on one of the existing pages for a user to manually sign in Acrobat.

The best solution atm is to append a signature page from an imported pdf that was prepared in Acrobat. Is there any way else to achieve this?

Thanks for your work on this library!

**Reply by ThomasHoevel** [reply]
Digital signatures are available in Preview 1. They look how you want them to look.
Feel free to try them out.

**Reply by ZephyrZiggurat** [reply]
Sorry, not blank as in visually but blank as in unfilled. I couldn't figure out a way to get a `PdfSignaturePlaceholderItem` added to the page except via this method: `public static DigitalSignatureHandler ForDocument(PdfDocument document, IDigitalSigner signer, DigitalSignatureOptions options)` which requires a certificate.

**Reply by ThomasHoevel** [reply]
The current user story is about creating a digital signature - and a certificate is required to achieve that.
Preparing a document to be signed by Adobe Acrobat is a different story. I don't know what is required to achieve that.

**Reply by Havunen** [reply]
@ThomasHoevel 
There is also a use case to create digital signature without certificate by using external device or service for the signing procedure.

1. Reserve space in the PDF document enough for the signature (something like 20kb)
2. Get the bytes from the pdf to be signed
3. Send the bytes to the external service / device to be signed
4. Write the raw bytes to the signature slot

**Reply by ThomasHoevel** [reply]
@Havunen AIUI, that use case is covered. Just implement IDigitalSigner to handle that.
You can also implement a dummy signer that just reserves memory for the signature, but does not add the actual signature.
All this is already covered.

@ZephyrZiggurat It depends on you whether your implementation of IDigitalSigner requires a certificate.

**Reply by Havunen** [reply]
@ThomasHoevel  

Is there some way to avoid calling `AddSignatureComponentsAsync` multiple times? 

https://github.com/empira/PDFsharp/blob/5fbf6ed14740bc4e16786816882d32e43af3ff5d/src/foundation/src/PDFsharp/src/PdfSharp/Pdf/PdfDocument.cs#L333-L334

I implemented custom signer type to reserve enough space in the PDF for the signature and then later I will somehow fill this information in. 

However it seems the Document will always add those signature components?

I'm planning the digital signature flow could go something like this:

1. Prepare the pdf document to be digitally signed, add the visual signatures and reserve space for signing.
2. Read the bytes out and send them to the client 1
3. client 1 will sign the document ( now the PDFSharp adds signing components again to the pdf document? )
4. later client 2 signs the document

In the signing phase, I would like to be able to edit the pdf only by writing the raw bytes to the digital signature block(s).
I thought implementing custom `DigitalSignatureHandler` but its not so easy because its static method is referenced from https://github.com/empira/PDFsharp/blob/5fbf6ed14740bc4e16786816882d32e43af3ff5d/src/foundation/src/PDFsharp/src/PdfSharp/Pdf.Signatures/PdfSignatureHandler.cs#L45-L48

**Reply by ThomasHoevel** [reply]
Implement your own IDigitalSigner, add bytes for the signature in step 1 without actually signing it, add digital signature in step 3 without using PDFsharp (locate the placeholder for the signature in the PDF and fill it).

**Reply by Havunen** [reply]
ah, true. I didn't think of adding the bytes without PDFSharp. yeah that should work 👍

**Reply by Havunen** [reply]
I noticed the bytes of `GetSignatureAsync(Stream stream)` contains all the bytes for the document to be signed, so that means only one signature can be done at a time, otherwise it would invalidate the other signature as the the signatures cover each other. 

Is there a method to create incremental version of the PDF so multiple signatures can be added one by one.

**Reply by Havunen** [reply]
https://github.com/empira/PDFsharp/pull/112

PR: https://github.com/empira/PDFsharp/pull/48

validator: https://ec.europa.eu/digital-building-blocks/DSS/webapp-demo/validation#SIGNATURE_Unit-Test_20241022-0703-CV

multiple signtures:
https://eideasy.com/pdf-pades-external-digital-signatures-using-etsi-cades-detached/

**Comment by diogopaulo** [comment]
When are you expecting to release PDFsharp 6.2? Thank you.

**Reply by mikethea1** [reply]
+1. We need the parsing fix. Looking forward to this release!

**Reply by ThomasHoevel** [reply]
> We need the parsing fix.

Which fix?

**Reply by mikethea1** [reply]
@ThomasHoevel sorry for being vague. Based on the release notes, I assume the relevant one is "Bug in parsing object referenced".

Specifically, [Sample file.pdf](https://github.com/user-attachments/files/17629221/Sample.file.pdf) throws PdfReaderException when opened under 6.1.1 but works in 6.2-preview-1.
